## Canadian Hospital Readmittance Challenge 

This notebook is made as a part of the Machine Learning (AI-511) project. It has been made by the following students - 

1. Siddharth Kothari (IMT2021019)
2. Sankalp Kothari (IMT2021028)
3. M Srinivasan (IMT2021058)


In [276]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.graph_objects as go
import xgboost as xgb
import optuna
from math import floor
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score,confusion_matrix,ConfusionMatrixDisplay
import re
pd.options.display.max_rows = 4000

In [277]:
df = pd.read_csv('./canadian-hospital-re-admittance-challenge/train.csv')
test_df = pd.read_csv('./canadian-hospital-re-admittance-challenge/test.csv')
df

enc_id  patient_id             race  gender       age     weight  \
0       88346340     2488608        Caucasian    Male   [60-70)        NaN   
1       92001408    52133202        Caucasian    Male   [70-80)  [100-125)   
2      169424316    40945509        Caucasian  Female   [70-80)        NaN   
3      272987082    38850777        Caucasian  Female   [50-60)        NaN   
4      150600612    72738225        Caucasian  Female   [80-90)        NaN   
...          ...         ...              ...     ...       ...        ...   
71231  198619164    85063725        Caucasian    Male   [70-80)        NaN   
71232  177404100    86244345        Caucasian    Male  [90-100)        NaN   
71233   50905206     5131368        Caucasian    Male   [70-80)        NaN   
71234  216431502    85969035         Hispanic    Male   [50-60)        NaN   
71235   32326944   112239351  AfricanAmerican  Female   [80-90)        NaN   

       admission_type_id  discharge_disposition_id  admission_source_id  \
0                      1                         2                    6   
1                      2                         6                    1   
2                      3                         2                    1   
3                      1                         1                    7   
4                      1                         6                    7   
...                  ...                       ...                  ...   
71231                  1                         1                    7   
71232                  1                         3                    7   
71233                  3                         6                    1   
71234                  1                         1                    4   
71235                  3                         6                    2   

       time_in_hospital  ... citoglipton insulin  glyburide-metformin  \
0                     3  ...          No  Steady                   No   
1                     7  ...          No      No                   No   
2                     7  ...          No      Up                   No   
3                     1  ...          No      No                   No   
4                     6  ...          No    Down                   No   
...                 ...  ...         ...     ...                  ...   
71231                 6  ...          No      No                   No   
71232                 5  ...          No      No                   No   
71233                 6  ...          No  Steady                   No   
71234                 4  ...          No  Steady                   No   
71235                 7  ...          No    Down                   No   

       glipizide-metformin  glimepiride-pioglitazone  metformin-rosiglitazone  \
0                       No                        No                       No   
1                       No                        No                       No   
2                       No                        No                       No   
3                       No                        No                       No   
4                       No                        No                       No   
...                    ...                       ...                      ...   
71231                   No                        No                       No   
71232                   No                        No                       No   
71233                   No                        No                       No   
71234                   No                        No                       No   
71235                   No                        No                       No   

       metformin-pioglitazone  change diabetesMed readmission_id  
0                          No      Ch         Yes              2  
1                          No      No         Yes              1  
2                          No      Ch         Yes              1  
3                          No      No         Yes     

### EDA and Preprocessing

In [278]:
df.columns

Index(['enc_id', 'patient_id', 'race', 'gender', 'age', 'weight',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'payer_code', 'medical_specialty',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1',
       'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult',
       'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide',
       'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmission_id'],
      dtype='object')

#### Null Removal

We first see the percentage of null values in each of the columns, to see which columns we can immediatelty drop.

In [279]:
percent_missing = df.isnull().sum() * 100 / len(df)

percentages_df = pd.DataFrame({
    'percent_missing': percent_missing
})

percentages_df

percent_missing
enc_id                           0.000000
patient_id                       0.000000
race                             2.275535
gender                           0.000000
age                              0.000000
weight                          96.841485
admission_type_id                0.000000
discharge_disposition_id         0.000000
admission_source_id              0.000000
time_in_hospital                 0.000000
payer_code                      39.555843
medical_specialty               49.034196
num_lab_procedures               0.000000
num_procedures                   0.000000
num_medications                  0.000000
number_outpatient                0.000000
number_emergency                 0.000000
number_inpatient                 0.000000
diag_1                           0.021057
diag_2                           0.342523
diag_3                           1.388343
number_diagnoses                 0.000000
max_glu_serum                   94.776517
A1Cresult                       83.323039
metformin                        0.000000
repaglinide                      0.000000
nateglinide                      0.000000
chlorpropamide                   0.000000
glimepiride                      0.000000
acetohexamide                    0.000000
glipizide                        0.000000
glyburide                        0.000000
tolbutamide                      0.000000
pioglitazone                     0.000000
rosiglitazone                    0.000000
acarbose                         0.000000
miglitol                         0.000000
troglitazone                     0.000000
tolazamide                       0.000000
examide                          0.000000
citoglipton                      0.000000
insulin                          0.000000
glyburide-metformin              0.000000
glipizide-metformin              0.000000
glimepiride-pioglitazone         0.000000
metformin-rosiglitazone          0.000000
metformin-pioglitazone           0.000000
change                           0.000000
diabetesMed                      0.000000
readmission_id                   0.000000

We see that the following columns have a very large number of null values - 
1. Weight (96.84)
2. A1Cresult (83.32)
3. max_glu_serum (94.77)


We thus try to see the distribution of the values in these columns to see whether any useful info can be gained.

In [280]:
df['max_glu_serum'].value_counts()

max_glu_serum
Norm    1790
>200    1034
>300     897
Name: count, dtype: int64

In [281]:
df['A1Cresult'].value_counts()

A1Cresult
>8      5715
Norm    3476
>7      2689
Name: count, dtype: int64

In [282]:
df['weight'].value_counts()

weight
[75-100)     944
[50-75)      643
[100-125)    421
[125-150)    103
[25-50)       69
[0-25)        34
[150-175)     27
[175-200)      7
>200           2
Name: count, dtype: int64

##### Dropping Columns

Since no useful info can be gained from these columns, we drop them altogether.

We also drop columns which do not seem to be relevant to the readmission of a patient, such as payer code.

In [283]:
df.drop(columns=['weight','payer_code','max_glu_serum','A1Cresult'], inplace=True)
test_df.drop(columns=['weight','payer_code','max_glu_serum','A1Cresult'],inplace=True)

df

enc_id  patient_id             race  gender       age  \
0       88346340     2488608        Caucasian    Male   [60-70)   
1       92001408    52133202        Caucasian    Male   [70-80)   
2      169424316    40945509        Caucasian  Female   [70-80)   
3      272987082    38850777        Caucasian  Female   [50-60)   
4      150600612    72738225        Caucasian  Female   [80-90)   
...          ...         ...              ...     ...       ...   
71231  198619164    85063725        Caucasian    Male   [70-80)   
71232  177404100    86244345        Caucasian    Male  [90-100)   
71233   50905206     5131368        Caucasian    Male   [70-80)   
71234  216431502    85969035         Hispanic    Male   [50-60)   
71235   32326944   112239351  AfricanAmerican  Female   [80-90)   

       admission_type_id  discharge_disposition_id  admission_source_id  \
0                      1                         2                    6   
1                      2                         6                    1   
2                      3                         2                    1   
3                      1                         1                    7   
4                      1                         6                    7   
...                  ...                       ...                  ...   
71231                  1                         1                    7   
71232                  1                         3                    7   
71233                  3                         6                    1   
71234                  1                         1                    4   
71235                  3                         6                    2   

       time_in_hospital       medical_specialty  ...  citoglipton  insulin  \
0                     3  Family/GeneralPractice  ...           No   Steady   
1                     7                     NaN  ...           No       No   
2                     7                     NaN  ...           No       Up   
3                     1                     NaN  ...           No       No   
4                     6                     NaN  ...           No     Down   
...                 ...                     ...  ...          ...      ...   
71231                 6                     NaN  ...           No       No   
71232                 5                     NaN  ...           No       No   
71233                 6              Cardiology  ...           No   Steady   
71234                 4        InternalMedicine  ...           No   Steady   
71235                 7                     NaN  ...           No     Down   

       glyburide-metformin  glipizide-metformin  glimepiride-pioglitazone  \
0                       No                   No                        No   
1                       No                   No                        No   
2                       No                   No                        No   
3                       No                   No                        No   
4                       No                   No                        No   
...                    ...                  ...                       ...   
71231                   No                   No                        No   
71232                   No                   No                        No   
71233                   No                   No                        No   
71234                   No                   No                        No   
71235                   No                   No                        No   

       metformin-rosiglitazone metformin-pioglitazone change diabetesMed  \
0                           No                     No     Ch         Yes   
1                           No                     No     No         Yes   
2                           No                     No     Ch         Yes   
3                           No                     No     No         Yes   
4                           No                     No     Ch  

##### Medical Specialty

For the medical specialty column, we do not directly drop it, as some specialists' patients may have more chance of being readmitted than others'. Hence to tackle null values there, we impute the null values with a new value - "No-admitting-Physician".

This assumption is supported by the image below

![Medical Specialty](./images/Medical%20Specialty.png)

In [284]:
df["medical_specialty"].fillna("No-Admitting-Physician", inplace=True)
test_df["medical_specialty"].fillna("No-Admitting-Physician", inplace=True)

In [285]:
percent_missing = df.isnull().sum() * 100 / len(df)

percentages_df = pd.DataFrame({
    'percent_missing': percent_missing
})

percentages_df

percent_missing
enc_id                           0.000000
patient_id                       0.000000
race                             2.275535
gender                           0.000000
age                              0.000000
admission_type_id                0.000000
discharge_disposition_id         0.000000
admission_source_id              0.000000
time_in_hospital                 0.000000
medical_specialty                0.000000
num_lab_procedures               0.000000
num_procedures                   0.000000
num_medications                  0.000000
number_outpatient                0.000000
number_emergency                 0.000000
number_inpatient                 0.000000
diag_1                           0.021057
diag_2                           0.342523
diag_3                           1.388343
number_diagnoses                 0.000000
metformin                        0.000000
repaglinide                      0.000000
nateglinide                      0.000000
chlorpropamide                   0.000000
glimepiride                      0.000000
acetohexamide                    0.000000
glipizide                        0.000000
glyburide                        0.000000
tolbutamide                      0.000000
pioglitazone                     0.000000
rosiglitazone                    0.000000
acarbose                         0.000000
miglitol                         0.000000
troglitazone                     0.000000
tolazamide                       0.000000
examide                          0.000000
citoglipton                      0.000000
insulin                          0.000000
glyburide-metformin              0.000000
glipizide-metformin              0.000000
glimepiride-pioglitazone         0.000000
metformin-rosiglitazone          0.000000
metformin-pioglitazone           0.000000
change                           0.000000
diabetesMed                      0.000000
readmission_id                   0.000000

##### Diag Groupings

The columns for diag_1, diag_2, diag_3 have a lot of categorical values, which we have grouped referring to the following website listing the groupings of ICD codes.

https://en.wikipedia.org/wiki/List_of_ICD-9_codes

We decided to drop the diag_1 column, and keep the other two, based on the following plots made in Tableau, which show the variation of the data with the columns. Readmission_id did not vary much with diag_1, hence we dropped it.

![Diag_1_group](./images/Diag_1_group.png)

![Diag_2_group](./images/Diag_2_group.png)

![Diag_3_group](./images/Diag_3_group.png)


In [286]:
diag_grouping_dict = {
    0 : [0],
    1 : range(1,140),
    2 : range(140,240),
    3 : range(240,280),
    4 : range(280,290),
    5 : range(290,320), 
    6 : range(320,390),
    7 : range(390,460),
    8 : range(460,520),
    9 : range(520,580),
    10: range(580,630),
    11: range(630,680),
    12: range(680,710),
    13: range(710,740),
    14: range(740,760),
    15: range(760-780),
    16: range(780,800),
    17: range(800,1000)
}

def diag_convert2(row):
    if str(row['diag_2'])[0] in ['E','V']:
        return 18
    else:
        for j in diag_grouping_dict.keys():
            if floor(float(row['diag_2'])) in diag_grouping_dict[j]:
                return j
            

def diag_convert3(row):
    if str(row['diag_3'])[0] in ['E','V']:
        return 18
    else:
        for j in diag_grouping_dict.keys():
            if floor(float(row['diag_3'])) in diag_grouping_dict[j]:
                return j

In [287]:
df.fillna("0", inplace=True)
test_df.fillna("0", inplace=True)

df.drop(columns=['diag_1'], inplace=True)
test_df.drop(columns=['diag_1'], inplace=True)

new_col = df.apply(diag_convert2, axis=1)
df.insert(loc = len(df.columns)-1, column = 'diag_2_new', value=new_col)
df.drop(columns=['diag_2'], inplace=True)

new_col = test_df.apply(diag_convert2, axis=1)
test_df.insert(loc = len(test_df.columns), column = 'diag_2_new', value=new_col)
test_df.drop(columns=['diag_2'], inplace=True)

new_col = df.apply(diag_convert3, axis=1)
df.insert(loc = len(df.columns)-1, column = 'diag_3_new', value=new_col)
df.drop(columns=['diag_3'], inplace=True)

new_col = test_df.apply(diag_convert3, axis=1)
test_df.insert(loc = len(test_df.columns), column = 'diag_3_new', value=new_col)
test_df.drop(columns=['diag_3'], inplace=True)

Another possibility we had considered was to drop diag_1, diag_2, diag_3 altogether, and instead replace it with a new column, which captures some of the information of the columns. However this did not bring about better results, so we dropped it.

In [288]:
# df.loc[df['diag_1'].notnull(), 'diag_1'] = 4
# df.loc[df['diag_2'].notnull(), 'diag_2'] = 2
# df.loc[df['diag_3'].notnull(), 'diag_3'] = 1

# df['diag_1'].fillna(0,inplace=True)
# df['diag_2'].fillna(0,inplace=True)
# df['diag_3'].fillna(0,inplace=True)

# test_df.loc[test_df['diag_1'].notnull(), 'diag_1'] = 4
# test_df.loc[test_df['diag_2'].notnull(), 'diag_2'] = 2
# test_df.loc[test_df['diag_3'].notnull(), 'diag_3'] = 1

# test_df['diag_1'].fillna(0,inplace=True)
# test_df['diag_2'].fillna(0,inplace=True)
# test_df['diag_3'].fillna(0,inplace=True)

# df.loc[:,'diag_1':'diag_3']

In [289]:
# new_col = df['diag_1']+df['diag_2']+df['diag_3']
# df.insert(loc = len(df.columns)-1, column = 'Number_of_Diagnosis', value=new_col)

# new_col = test_df['diag_1']+test_df['diag_2']+test_df['diag_3']
# test_df.insert(loc = len(test_df.columns), column = 'Number_of_Diagnosis', value=new_col)

In [290]:
# df.drop(columns=['diag_1','diag_2','diag_3'], inplace=True)
# test_df.drop(columns=['diag_1','diag_2','diag_3'], inplace=True)

##### Race and Gender

Finally, using the plots below, we realised that the race and gender of the person did not affect the values of readmission_id observed, hence we felt it better to drop these two columns.

Our claims are supported by the charts below.


![Gender](./images/Gender.png)

![Race](./images/Race.png)

In [291]:
df.drop(columns=['race','gender'], inplace=True)
test_df.drop(columns=['race','gender'], inplace=True)
df

enc_id  patient_id       age  admission_type_id  \
0       88346340     2488608   [60-70)                  1   
1       92001408    52133202   [70-80)                  2   
2      169424316    40945509   [70-80)                  3   
3      272987082    38850777   [50-60)                  1   
4      150600612    72738225   [80-90)                  1   
...          ...         ...       ...                ...   
71231  198619164    85063725   [70-80)                  1   
71232  177404100    86244345  [90-100)                  1   
71233   50905206     5131368   [70-80)                  3   
71234  216431502    85969035   [50-60)                  1   
71235   32326944   112239351   [80-90)                  3   

       discharge_disposition_id  admission_source_id  time_in_hospital  \
0                             2                    6                 3   
1                             6                    1                 7   
2                             2                    1                 7   
3                             1                    7                 1   
4                             6                    7                 6   
...                         ...                  ...               ...   
71231                         1                    7                 6   
71232                         3                    7                 5   
71233                         6                    1                 6   
71234                         1                    4                 4   
71235                         6                    2                 7   

            medical_specialty  num_lab_procedures  num_procedures  ...  \
0      Family/GeneralPractice                  54               3  ...   
1      No-Admitting-Physician                  50               6  ...   
2      No-Admitting-Physician                  49               4  ...   
3      No-Admitting-Physician                   1               1  ...   
4      No-Admitting-Physician                  58               2  ...   
...                       ...                 ...             ...  ...   
71231  No-Admitting-Physician                  61               0  ...   
71232  No-Admitting-Physician                  36               0  ...   
71233              Cardiology                  62               6  ...   
71234        InternalMedicine                  29               1  ...   
71235  No-Admitting-Physician                  35               2  ...   

       glyburide-metformin  glipizide-metformin  glimepiride-pioglitazone  \
0                       No                   No                        No   
1                       No                   No                        No   
2                       No                   No                        No   
3                       No                   No                        No   
4                       No                   No                        No   
...                    ...                  ...                       ...   
71231                   No                   No                        No   
71232                   No                   No                        No   
71233                   No                   No                        No   
71234                   No                   No                        No   
71235                   No                   No                        No   

       metformin-rosiglitazone  metformin-pioglitazone change diabetesMed  \
0                           No                      No     Ch         Yes   
1                           No                      No     No         Yes   
2                           No                      No     Ch         Yes   
3                           No                      No     No         Yes   
4                           No                      No     Ch         Yes   
...                        ...                     ...    ...         ...   
71231                       No     

#### Engineering new features

We also made some new features which we thought would better capture the data. 

1. Admission_type_id_new - here we grouped the various values in the columns as given in admission_grouping_dict to reduce the number of categories.
2. Discharge_type_id_new - similar as previous. We grouped them based on which categories broadly fell into the same umbrella according to our understanding. 
3. Admission_source_id_new - similar as previous.

The mapping of the number to the category is as follows -

##### Admission type id
| admission_type_id | description |
| ----------------- | ----------- |
| 1 | Emergency |
| 2 | Urgent |
| 3 | Elective |
| 4 | Newborn |
| 5 | Not Available |
| 6 | NULL |
| 7 | Trauma Center |
| 8 | Not Mapped |

##### Discharge Disposition id

| discharge_disposition_id | description |
| ------------------------ | ----------- |
| 1 | Discharged to home |
| 2 | Discharged/transferred to another short term hospital |
| 3 | Discharged/transferred to SNF(Skilled Nursing Facility). |
| 4 | Discharged/transferred to ICF(Intermediate Care Facility). |
| 5 | Discharged/transferred to another type of inpatient care institution |
| 6 | Discharged/transferred to home with home health service |
| 7 | Left AMA (American Medical Association) |
| 8 | Discharged/transferred to home under care of Home IV provider |
| 9 | Admitted as an inpatient to this hospital |
| 10 | Neonate discharged to another hospital for neonatal aftercare |
| 11 | Expired |
| 12 | Still patient or expected to return for outpatient services |
| 13 | Hospice(Nursing Home for ill/old) / home |
| 14 | Hospice / medical facility |
| 15 | Discharged/transferred within this institution to Medicare approved swing bed |
| 16 | Discharged/transferred/referred another institution for outpatient services |
| 17 | Discharged/transferred/referred to this institution for outpatient services |
| 18 | NULL |
| 19 | "Expired at home. Medicaid only, hospice." |
| 20 | "Expired in a medical facility. Medicaid only, hospice." |
| 21 | "Expired, place unknown. Medicaid only, hospice." |
| 22 | Discharged/transferred to another rehab fac including rehab units of a hospital. |
| 23 | Discharged/transferred to a long term care hospital. |
| 24 | Discharged/transferred to a nursing facility certified under Medicaid but not certified under Medicare. |
| 25 | Not Mapped |
| 26 | Unknown/Invalid |
| 30 | Discharged/transferred to another Type of Health Care Institution not Defined Elsewhere |
| 27 | Discharged/transferred to a federal health care facility. |
| 28 | Discharged/transferred/referred to a psychiatric hospital of psychiatric distinct part unit of a hospital |
| 29 | Discharged/transferred to a Critical Access Hospital (CAH). |


##### Admission source id
| admission_source_id | description |
| ------------------- | ----------- |
| 1  | Physician Referral |
| 2  | Clinic Referral |
| 3  | HMO Referral |
| 4  | Transfer from a hospital |
| 5  | Transfer from a Skilled Nursing Facility (SNF) | 
| 6  | Transfer from another health care facility | 
| 7  | Emergency Room | 
| 8  | Court/Law Enforcement |
| 9  | Not Available |
| 10 | Transfer from critial access hospital |
| 11 | Normal Delivery |
| 12 | Premature Delivery |
| 13 | Sick Baby |
| 14 | Extramural Birth |
| 15 | Not Available |
| 17 | NULL |
| 18 | Transfer From Another Home Health Agency |
| 19 | Readmission to Same Home Health Agency |
| 20 | Not Mapped |
| 21 | Unknown/Invalid |
| 22 | Transfer from hospital inpt/same fac reslt in a sep claim |
| 23 | Born inside this hospital |
| 24 | Born outside this hospital |
| 25 | Transfer from Ambulatory Surgery Center |
| 26 | Transfer from Hospice |

In [292]:
admission_grouping_dict = {
    1 : [1],
    2 : [2],
    3 : [3],
    4 : [4],
    5 : [5,6,8], 
    6 : [7]
}

def admission_group(row):
    for j in admission_grouping_dict.keys():
        if row['admission_type_id'] in admission_grouping_dict[j]:
            return j

new_col = df.apply(admission_group, axis=1)
df.insert(loc = len(df.columns)-1, column = 'admission_type_id_new', value=new_col)
df.drop(columns=['admission_type_id'], inplace=True)

new_col = test_df.apply(admission_group, axis=1)
test_df.insert(loc = len(test_df.columns), column = 'admission_type_id_new', value=new_col)
test_df.drop(columns=['admission_type_id'], inplace=True)

df['admission_type_id_new'].value_counts()

admission_type_id_new
1    37831
3    13188
2    12979
5     7223
4        8
6        7
Name: count, dtype: int64

In [293]:
discharge_grouping_dict = {
    1 : [11,19,20,21],
    2 : [18,25,26],
    3 : [7],
    4 : [1,6,8,13,14],
    5 : [2,3,4,5,10,16,22,23,24,30,27,28,29],
    6 : [9,12,15,17]
}

def discharge_group(row):
    for j in discharge_grouping_dict.keys():
        if row['discharge_disposition_id'] in discharge_grouping_dict[j]:
            return j
        
new_col = df.apply(discharge_group, axis=1)
df.insert(loc = len(df.columns)-1, column = 'discharge_type_id_new', value=new_col)
df.drop(columns=['discharge_disposition_id'], inplace=True)

new_col = test_df.apply(discharge_group, axis=1)
test_df.insert(loc = len(test_df.columns), column = 'discharge_type_id_new', value=new_col)
test_df.drop(columns=['discharge_disposition_id'], inplace=True)

df['discharge_type_id_new'].value_counts()

discharge_type_id_new
4    51763
5    14489
2     3319
1     1160
3      444
6       61
Name: count, dtype: int64

In [294]:
source_grouping_dict = {
    1 : [4,5,6,10,18,22,25,26],
    2 : [1,2,3],
    3 : [11,12,13,14],
    4 : [9,15,17,20,21],
    5 : [7],
    6 : [8]
}

def source_group(row):
    for j in source_grouping_dict.keys():
        if row['admission_source_id'] in source_grouping_dict[j]:
            return j

new_col = df.apply(source_group, axis=1)
df.insert(loc = len(df.columns)-1, column = 'admission_source_id_new', value=new_col)
df.drop(columns=['admission_source_id'], inplace=True)

new_col = test_df.apply(source_group, axis=1)
test_df.insert(loc = len(test_df.columns), column = 'admission_source_id_new', value=new_col)
test_df.drop(columns=['admission_source_id'], inplace=True)

df['admission_source_id_new'].value_counts()

admission_source_id_new
5    40268
2    21610
4     4877
1     4465
6       11
3        5
Name: count, dtype: int64

4. Service_utilization - we felt that the total number of visits by a person (emergency + inpatient + outpatient) can be a good indicator of whether the person will be readmitted or not. Hence we created this new feature. 

In [295]:
new_col = df['number_outpatient'] + df['number_emergency'] + df['number_inpatient']
df.insert(loc = len(df.columns)-1, column = 'service_utilization', value=new_col)

new_col = test_df['number_outpatient'] + test_df['number_emergency'] + test_df['number_inpatient']
test_df.insert(loc = len(test_df.columns), column = 'service_utilization', value=new_col)

df

enc_id  patient_id       age  time_in_hospital  \
0       88346340     2488608   [60-70)                 3   
1       92001408    52133202   [70-80)                 7   
2      169424316    40945509   [70-80)                 7   
3      272987082    38850777   [50-60)                 1   
4      150600612    72738225   [80-90)                 6   
...          ...         ...       ...               ...   
71231  198619164    85063725   [70-80)                 6   
71232  177404100    86244345  [90-100)                 5   
71233   50905206     5131368   [70-80)                 6   
71234  216431502    85969035   [50-60)                 4   
71235   32326944   112239351   [80-90)                 7   

            medical_specialty  num_lab_procedures  num_procedures  \
0      Family/GeneralPractice                  54               3   
1      No-Admitting-Physician                  50               6   
2      No-Admitting-Physician                  49               4   
3      No-Admitting-Physician                   1               1   
4      No-Admitting-Physician                  58               2   
...                       ...                 ...             ...   
71231  No-Admitting-Physician                  61               0   
71232  No-Admitting-Physician                  36               0   
71233              Cardiology                  62               6   
71234        InternalMedicine                  29               1   
71235  No-Admitting-Physician                  35               2   

       num_medications  number_outpatient  number_emergency  ...  \
0                   10                  0                 0  ...   
1                   35                  2                 1  ...   
2                   20                  2                 0  ...   
3                   15                  0                 0  ...   
4                   26                  1                 0  ...   
...                ...                ...               ...  ...   
71231               14                  0                 0  ...   
71232               12                  2                 0  ...   
71233               28                  0                 0  ...   
71234               11                  0                 0  ...   
71235               37                  0                 2  ...   

       metformin-pioglitazone  change diabetesMed diag_2_new diag_3_new  \
0                          No      Ch         Yes          7          7   
1                          No      No         Yes          7          7   
2                          No      Ch         Yes          8         17   
3                          No      No         Yes         10          3   
4                          No      Ch         Yes         13         12   
...                       ...     ...         ...        ...        ...   
71231                      No      No         Yes          3          7   
71232                      No      No          No          8          8   
71233                      No      Ch         Yes          4          3   
71234                      No      No         Yes          9          4   
71235                      No      Ch         Yes         12          7   

      admission_type_id_new discharge_type_id_new admission_source_id_new  \
0                         1                     5                       1   
1                         2                     4                       2   
2                         3                     5                       2   
3                         1                     4                       5   
4                         1                     4                       5   
...                     ...                   ...                     ...   
71231                     1                     4                       5   
71232                     1                     5                       5   
71233                     3                     4

5. Age - age is made as a categorical variable here, and we felt it should be numeric. Hence for each range (eg. 40-50) we replace the value with the median of the range, to make it numeric.

In [296]:
def age_converter(row):
    if row['age'] == '[0-10)':
        return 5
    elif row['age'] == '[10-20)':
        return 15
    elif row['age'] == '[20-30)':
        return 25
    elif row['age'] == '[30-40)':
        return 35
    elif row['age'] == '[40-50)':
        return 45
    elif row['age'] == '[50-60)':
        return 55
    elif row['age'] == '[60-70)':
        return 65
    elif row['age'] == '[70-80)':
        return 75
    elif row['age'] == '[80-90)':
        return 85
    elif row['age'] == '[90-100)':
        return 95

new_col = df.apply(age_converter, axis=1)
df['age'] = new_col

new_col = test_df.apply(age_converter, axis=1)
test_df['age'] = new_col

df['age'].value_counts()

age
75    18179
65    15801
55    12080
85    12037
45     6785
35     2650
95     1940
25     1165
15      495
5       104
Name: count, dtype: int64

##### Patient visit count

We felt the number of entries of admittance of a patient (indicated by the number of records of a given patient id) can be a useful measure of whether that person has been readmitted. Hence we count the number of records of the patient and make a new column "patient_id_new" for the same. We then drop enc_id and patient_id as those are anyways not useful features, and as such can't tell anything about the readmittance of a patient.

In [297]:
patient = df['patient_id'].value_counts()
patient = patient.add(test_df['patient_id'].value_counts(),fill_value=0)
patient = patient.astype('int64')

def id_convertor(row):
    return patient[row['patient_id']]

new_col = df.apply(id_convertor, axis=1)
df.insert(loc = len(df.columns)-1, column = 'patient_id_new', value=new_col)
df.drop(columns=['patient_id'],inplace=True)
df.drop(columns=['enc_id'], inplace=True)

new_col = test_df.apply(id_convertor, axis=1)
test_df.insert(loc = len(test_df.columns), column = 'patient_id_new', value=new_col)
test_df.drop(columns=['patient_id'], inplace=True)
test_df.drop(columns=['enc_id'], inplace=True)

##### Drug manipulations

We realised that for a lot of patients, many of the drugs mentioned are either not administered, or their dosage has not been changed. Hence we first count the number for each of the drugs.

In [298]:
for col in df.loc[:,'metformin':'diabetesMed']:
    med_groups = df.groupby(by=[col])
    print(med_groups.count().iloc[:, 0])

metformin
Down        392
No        57223
Steady    12885
Up          736
Name: age, dtype: int64
repaglinide
Down         33
No        70145
Steady      980
Up           78
Name: age, dtype: int64
nateglinide
Down          8
No        70770
Steady      444
Up           14
Name: age, dtype: int64
chlorpropamide
No        71170
Steady       62
Up            4
Name: age, dtype: int64
glimepiride
Down        131
No        67558
Steady     3308
Up          239
Name: age, dtype: int64
acetohexamide
No        71235
Steady        1
Name: age, dtype: int64
glipizide
Down        376
No        62301
Steady     8011
Up          548
Name: age, dtype: int64
glyburide
Down        389
No        63713
Steady     6577
Up          557
Name: age, dtype: int64
tolbutamide
No        71221
Steady       15
Name: age, dtype: int64
pioglitazone
Down         83
No        66074
Steady     4910
Up          169
Name: age, dtype: int64
rosiglitazone
Down         67
No        66740
Steady     4303
Up          126
Na

We then see that the following drugs have a very small number of records where the dosage of a drug has been changed, and as such it won't contribute much to the model. Hence we remove the following drugs - 
1. chlorpropamide
2. tolbutamide
3. miglitol
4. acarbose
5. tolazamide
6. acetohexamide
7. troglitazone
8. examide
9. citoglipton
10. glipizide-metformin
11. glimepiride-pioglitazone
12. metformin-rosiglitazone
13. metformin-pioglitazone
14. glyburide-metformin

In [299]:
df.drop(columns=['chlorpropamide', 'tolbutamide', 'miglitol', 'acarbose', 'tolazamide', 'acetohexamide', 'troglitazone', 'examide', 'citoglipton', 'glipizide-metformin', 'glimepiride-pioglitazone', 'metformin-rosiglitazone', 'metformin-pioglitazone', 'glyburide-metformin'], inplace=True)
test_df.drop(columns=['chlorpropamide', 'tolbutamide', 'miglitol', 'acarbose', 'tolazamide', 'acetohexamide', 'troglitazone', 'examide', 'citoglipton', 'glipizide-metformin', 'glimepiride-pioglitazone', 'metformin-rosiglitazone', 'metformin-pioglitazone', 'glyburide-metformin'], inplace=True)
df.columns.size

29

For the remaining drugs, we then count the number of "changes" of drug dosages, i.e. for a person, how many of the dosages either went up or down. We make this as a new feature - "changes", and then drop all the remaining drug columns.

In [300]:
def count_changes(row):
    count =0
    for col in ['metformin','repaglinide','nateglinide','glimepiride','glipizide','glyburide','pioglitazone','rosiglitazone','insulin']:
        if(row[col]=='Up' or row[col]=='Down'):
            count+=1
    return count

new_col = df.apply(count_changes, axis=1)
df.insert(loc = len(df.columns)-1, column = 'changes', value=new_col)

new_col = test_df.apply(count_changes, axis=1)
test_df.insert(loc = len(test_df.columns), column = 'changes', value=new_col)

In [301]:
df.drop(columns=['metformin','repaglinide','nateglinide','glimepiride','glipizide','glyburide','pioglitazone','rosiglitazone','insulin','change'], inplace=True)
test_df.drop(columns=['metformin','repaglinide','nateglinide','glimepiride','glipizide','glyburide','pioglitazone','rosiglitazone','insulin','change'], inplace=True)

df.columns.size

20

Based on the following plots, we also considered dropping the following columns, as they didn't seem to affect the distribution of readmission_id.

1. num_procedures
2. time_in_hospital

![num_procedures](./images/num_procedures.png)

![time_in_hospital](./images/time_in_hospital.png)

But that led to a decrease in accuracy, so we didn't end up doing this.

In [302]:
# df.drop(columns=['num_procedures','time_in_hospital'])
# test_df.drop(columns=['num_procedures','time_in_hospital'])

#### Data Visualization 

We plotted the following plots to get a better idea as to how the data affects readmission_id.
 
1. Scatterplot matrix between the various numeric features. The code for it has been commented out below, as it takes considerable amount of time to run. The plot has been downloaded and displayed below.
2. Covariance plot between the various numeric features and readmission_id. The code for the same is available below, and it is also commented out.  
3. Stacked bar charts for categorical variables, similar to the ones shown previously. They have been attached as well.


![Splom](./images/scatterplot_matrix.png)

![Covariance](./images/covariance.png)

![Admission_type_id](./images/Admission_type_id.png)

![Discharge_disposition_id](./images/Discharge_disposition_id.png)

![Admission_source_id](./images/Admission_source_id.png)

The other plots can be found in the images directory. We weren't able to infer much from them, and hence they weren't included in the report.

In [303]:
# plt.figure(figsize=(20,20))
# corr = df.loc[:, ["time_in_hospital","num_lab_procedures","num_procedures","num_medications","number_outpatient","number_emergency","number_inpatient","number_diagnoses","changes", "readmission_id"]].corr()
# sns.heatmap(corr,annot=True)

In [304]:
# index_vals = df['readmission_id'].astype('category').cat.codes

# dimensions = []

# cols = ['num_lab_procedures','num_procedures','num_medications','time_in_hospital','service_utilization','number_outpatient', 'number_emergency', 'number_inpatient', 'number_diagnoses', 'changes']

# for i in cols:
#     d = dict(label=i, values=df[i])
#     dimensions.append(d)

# fig = go.Figure(data=go.Splom(
#      dimensions=dimensions,
#      diagonal_visible=False,
#      text=df['readmission_id'],
#      marker=dict(color=index_vals,
#                line_color='white', line_width=0.5)
# ))

# fig.update_layout(
#     title='Readmission id',
#     dragmode='select',
#     width=1500,
#     height=1500,
#     hovermode='closest',
# )

# fig.show()


#### Model fitting 

We tried out various combinations of models for the data. Before that, we first carried out a few steps - 

1. We standard scaled the numerical columns.
2. We one-hot encoded all the categorical columns with 3 or more categories. This led to certain columns which were there in the test data but not in the training data (as some categories appeared only in the test data), and some others which were only available in the training data.
3. To handle this, we do the following - add all columns from the train data not present in the test data with all values as zero, while the columns in the test data not available in the test data are simply dropped. We then sort the columns based on name to ensure correct ordering in training and testing data.
4. We then do a train test split with the test size as 0.2, and proceed to fit the models.
5. We also tried introducing polynomial features, but it decreased the performance of the model, so we removed it.

In [305]:
df.columns

Index(['age', 'time_in_hospital', 'medical_specialty', 'num_lab_procedures',
       'num_procedures', 'num_medications', 'number_outpatient',
       'number_emergency', 'number_inpatient', 'number_diagnoses',
       'diabetesMed', 'diag_2_new', 'diag_3_new', 'admission_type_id_new',
       'discharge_type_id_new', 'admission_source_id_new',
       'service_utilization', 'patient_id_new', 'changes', 'readmission_id'],
      dtype='object')

In [306]:
test_df.columns

Index(['age', 'time_in_hospital', 'medical_specialty', 'num_lab_procedures',
       'num_procedures', 'num_medications', 'number_outpatient',
       'number_emergency', 'number_inpatient', 'number_diagnoses',
       'diabetesMed', 'diag_2_new', 'diag_3_new', 'admission_type_id_new',
       'discharge_type_id_new', 'admission_source_id_new',
       'service_utilization', 'patient_id_new', 'changes'],
      dtype='object')

In [307]:
input = df.loc[:, "age":"changes"]
labels = df.loc[:, "readmission_id"]
input.columns

Index(['age', 'time_in_hospital', 'medical_specialty', 'num_lab_procedures',
       'num_procedures', 'num_medications', 'number_outpatient',
       'number_emergency', 'number_inpatient', 'number_diagnoses',
       'diabetesMed', 'diag_2_new', 'diag_3_new', 'admission_type_id_new',
       'discharge_type_id_new', 'admission_source_id_new',
       'service_utilization', 'patient_id_new', 'changes'],
      dtype='object')

In [308]:
input_encoded = pd.get_dummies(input, columns=['medical_specialty','admission_type_id_new',
                    'discharge_type_id_new','admission_source_id_new','diag_2_new','diag_3_new'])

input_encoded['diabetesMed'] = np.where(input_encoded['diabetesMed']=='Yes',1,0)

print(input_encoded.columns)

Index(['age', 'time_in_hospital', 'num_lab_procedures', 'num_procedures',
       'num_medications', 'number_outpatient', 'number_emergency',
       'number_inpatient', 'number_diagnoses', 'diabetesMed',
       ...
       'diag_3_new_8', 'diag_3_new_9', 'diag_3_new_10', 'diag_3_new_11',
       'diag_3_new_12', 'diag_3_new_13', 'diag_3_new_14', 'diag_3_new_16',
       'diag_3_new_17', 'diag_3_new_18'],
      dtype='object', length=136)


In [309]:
test_encoded = pd.get_dummies(test_df, columns=['medical_specialty','admission_type_id_new', 
                    'discharge_type_id_new','admission_source_id_new','diag_2_new','diag_3_new'])

test_encoded['diabetesMed'] = np.where(test_encoded['diabetesMed']=='Yes',1,0)


print(test_encoded.columns)

Index(['age', 'time_in_hospital', 'num_lab_procedures', 'num_procedures',
       'num_medications', 'number_outpatient', 'number_emergency',
       'number_inpatient', 'number_diagnoses', 'diabetesMed',
       ...
       'diag_3_new_8', 'diag_3_new_9', 'diag_3_new_10', 'diag_3_new_11',
       'diag_3_new_12', 'diag_3_new_13', 'diag_3_new_14', 'diag_3_new_16',
       'diag_3_new_17', 'diag_3_new_18'],
      dtype='object', length=130)


In [310]:
for i in input_encoded.columns:
    if i not in test_encoded.columns:
       test_encoded[i] = 0

for i in test_encoded.columns:
    if i not in input_encoded.columns:
       test_encoded.drop(columns=[i], inplace=True)

input_encoded.sort_index(axis=1, inplace=True)
test_encoded.sort_index(axis=1, inplace=True)

In [311]:
print(input_encoded.columns)
print(test_encoded.columns)

Index(['admission_source_id_new_1', 'admission_source_id_new_2',
       'admission_source_id_new_3', 'admission_source_id_new_4',
       'admission_source_id_new_5', 'admission_source_id_new_6',
       'admission_type_id_new_1', 'admission_type_id_new_2',
       'admission_type_id_new_3', 'admission_type_id_new_4',
       ...
       'num_lab_procedures', 'num_medications', 'num_procedures',
       'number_diagnoses', 'number_emergency', 'number_inpatient',
       'number_outpatient', 'patient_id_new', 'service_utilization',
       'time_in_hospital'],
      dtype='object', length=136)
Index(['admission_source_id_new_1', 'admission_source_id_new_2',
       'admission_source_id_new_3', 'admission_source_id_new_4',
       'admission_source_id_new_5', 'admission_source_id_new_6',
       'admission_type_id_new_1', 'admission_type_id_new_2',
       'admission_type_id_new_3', 'admission_type_id_new_4',
       ...
       'num_lab_procedures', 'num_medications', 'num_procedures',
       'number

In [312]:
# scaler = StandardScaler()
# input_encoded[['patient_id_new','age','num_lab_procedures','number_outpatient', 'number_emergency', 'number_inpatient', 'number_diagnoses', 'changes','num_procedures','num_medications','time_in_hospital']] = scaler.fit_transform(input_encoded[['patient_id_new','age','num_lab_procedures','number_outpatient', 'number_emergency', 'number_inpatient', 'number_diagnoses', 'changes','num_procedures','num_medications','time_in_hospital']].to_numpy())
# test_encoded[['patient_id_new','age','num_lab_procedures','number_outpatient', 'number_emergency', 'number_inpatient', 'number_diagnoses', 'changes','num_procedures','num_medications','time_in_hospital']] = scaler.fit_transform(test_encoded[['patient_id_new','age','num_lab_procedures','number_outpatient', 'number_emergency', 'number_inpatient', 'number_diagnoses', 'changes','num_procedures','num_medications','time_in_hospital']].to_numpy())

In [313]:
# preprocessor = ColumnTransformer(
#     transformers=[
#         ('transformer', PolynomialFeatures(degree=2, include_bias=False), ['time_in_hospital', 'num_lab_procedures', 'num_procedures', 'num_medications', 'number_outpatient', 'number_emergency', 'number_inpatient', 'number_diagnoses']),
#     ],
#     remainder='passthrough'
# )

# input_encoded = preprocessor.fit_transform(input_encoded)
# test_encoded = preprocessor.transform(test_encoded)

input_encoded['readmission_id'] = labels
input_encoded

admission_source_id_new_1  admission_source_id_new_2  \
0                           True                      False   
1                          False                       True   
2                          False                       True   
3                          False                      False   
4                          False                      False   
...                          ...                        ...   
71231                      False                      False   
71232                      False                      False   
71233                      False                       True   
71234                       True                      False   
71235                      False                       True   

       admission_source_id_new_3  admission_source_id_new_4  \
0                          False                      False   
1                          False                      False   
2                          False                      False   
3                          False                      False   
4                          False                      False   
...                          ...                        ...   
71231                      False                      False   
71232                      False                      False   
71233                      False                      False   
71234                      False                      False   
71235                      False                      False   

       admission_source_id_new_5  admission_source_id_new_6  \
0                          False                      False   
1                          False                      False   
2                          False                      False   
3                           True                      False   
4                           True                      False   
...                          ...                        ...   
71231                       True                      False   
71232                       True                      False   
71233                      False                      False   
71234                      False                      False   
71235                      False                      False   

       admission_type_id_new_1  admission_type_id_new_2  \
0                         True                    False   
1                        False                     True   
2                        False                    False   
3                         True                    False   
4                         True                    False   
...                        ...                      ...   
71231                     True                    False   
71232                     True                    False   
71233                    False                    False   
71234                     True                    False   
71235                    False                    False   

       admission_type_id_new_3  admission_type_id_new_4  ...  num_medications  \
0                        False                    False  ...               10   
1                        False                    False  ...               35   
2                         True                    False  ...               20   
3                        False                    False  ...               15   
4                        False                    False  ...               26   
...                        ...                      ...  ...              ...   
71231                    False                    False  ...               14   
71232                    False                    False  ...               12   
71233                     True                    False  ...               28   
71234                    False                    False  ...               11   
71235                     True                    False  ...               37   

       num_procedures  number_diagnoses  number_emerge

In [252]:
input_encoded.to_csv("processed_data.csv", index=False)
test_encoded.to_csv("test_data.csv", index=False)

In [ ]:
X_train,X_test,Y_train,Y_test = train_test_split(input_encoded, labels, test_size=0.2, random_state=42)

We then proceed to try a number of models. We used optuna for hyperparameter tuning, and the best cross validation accuracy scores we got for 5 models that we trained are as follows - 

| Model | CV Score |
| ----- | -------- |
| Logistic Regression | 0.6647248736664795 | 
| Gaussian Naive Bayes | 0.16282987085906794 | 
| Decision Tree | 0.6530039303761932 |
| Random Forest | 0.7114682762492981 |
| XGBoost (Gradient Boosting) | 0.7120297585626053 | 

Based on this we chose xgboost as the best model, and used it to make predictions on the test data. 

In [ ]:
# def objective(trial):
#     criterion = trial.suggest_categorical("criterion", ["gini", "entropy"])
#     max_depth = trial.suggest_int("max_depth", 2, 32, log=True)
#     n_estimators = trial.suggest_int("n_estimators", 100,500)
#     random_state = trial.suggest_int("random_state",42,42)
#     rf = RandomForestClassifier(criterion =criterion,
#             max_depth=max_depth, 
#             n_estimators=n_estimators,
#             random_state=random_state
#         )
#     X_train,X_test,Y_train,Y_test = train_test_split(input_encoded, labels, test_size=0.2, random_state=42)
#     rf.fit(X_train,Y_train)
#     y_pred = rf.predict(X_test)
#     score = accuracy_score(y_pred, Y_test)
#     return score


# study = optuna.create_study(direction="maximize")
# study.optimize(objective, n_trials=15)


# def objective2(trial):
#     # data, target = sklearn.datasets.load_breast_cancer(return_X_y=True)
#     X_train,X_test,Y_train,Y_test = train_test_split(input_encoded, labels, test_size=0.3, random_state=42)
#     regex = re.compile(r"\[|\]|<", re.IGNORECASE)
#     dict ={0:1.45,2:1,1:1.4}
#     X_train.columns = [regex.sub("_", col) if any(x in str(col) for x in set(('[', ']', '<'))) else col for col in X_train.columns.values]
#     X_test.columns = [regex.sub("_", col) if any(x in str(col) for x in set(('[', ']', '<'))) else col for col in X_test.columns.values]

#     max_depth = trial.suggest_int("max_depth", 3, 10)
#     n_estimators = trial.suggest_int("n_estimators", 200,500)
#     learning_rate = trial.suggest_int("learning_rate",0,1)
#     # gamma = trial.suggest_int("gamma",0,5)
#     reg_lambda = trial.suggest_int("reg_lambda",0,5)
#     class_weight = trial.suggest_int("class_weight",0,3)
#     rf = xgb.XGBClassifier(
#             max_depth=max_depth, 
#             n_estimators=n_estimators,
#             learning_rate=learning_rate,
#             reg_lambda=reg_lambda,
#             class_weight = class_weight
#         )
#     rf.fit(X_train,Y_train)
#     preds = rf.predict(X_test)
#     pred_labels = np.rint(preds)
#     accuracy = accuracy_score(Y_test, pred_labels)
#     return accuracy

# study = optuna.create_study(direction="maximize")
# study.optimize(objective2, n_trials=15)


In [ ]:
# lr = LogisticRegression(random_state=42, multi_class="multinomial")
# lr.fit(X_train,Y_train)

# y_pred = lr.predict(X_test)
# print(accuracy_score(y_pred, Y_test))

# nb = GaussianNB()
# nb.fit(X_train,Y_train)

# y_pred = nb.predict(X_test)
# print(accuracy_score(y_pred, Y_test))

# tree = DecisionTreeClassifier(max_depth=20,random_state=42)
# tree.fit(X_train,Y_train)

# y_pred = tree.predict(X_test)
# print(accuracy_score(y_pred, Y_test))

# rf = RandomForestClassifier(random_state=42, criterion='entropy', max_depth=30, n_estimators=440)
# rf.fit(X_train,Y_train)

# y_pred = rf.predict(X_test)
# print(accuracy_score(y_pred, Y_test))

In [ ]:
regex = re.compile(r"\[|\]|<", re.IGNORECASE)
 
X_train.columns = [regex.sub("_", col) if any(x in str(col) for x in set(('[', ']', '<'))) else col for col in X_train.columns.values]
X_test.columns = [regex.sub("_", col) if any(x in str(col) for x in set(('[', ']', '<'))) else col for col in X_test.columns.values]
test_encoded.columns = [regex.sub("_", col) if any(x in str(col) for x in set(('[', ']', '<'))) else col for col in test_encoded.columns.values]


xgb = xgb.XGBClassifier(max_depth=3,n_estimators=208,learning_rate=1,reg_lambda=3,class_weight=2)
xgb.fit(X_train,Y_train)

y_pred = xgb.predict(X_test)
print(accuracy_score(y_pred, Y_test))

[11:36:34] WARNING: /croot/xgboost-split_1675457761144/work/src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

0.7120297585626053


In [ ]:
test_Y = xgb.predict(test_encoded)

df_output = pd.read_csv("./sample_submission.csv")
df_output["readmission_id"] = test_Y
df_output.to_csv("submission.csv", index=False)

We then finally plot the confusion matrix for the same.

In [ ]:
# cm = confusion_matrix(Y_test, y_pred, labels=xgb.classes_)
# disp = ConfusionMatrixDisplay(confusion_matrix=cm,display_labels=xgb.classes_)
# disp.plot()
# plt.show()
# print(accuracy_score(y_pred, Y_test))

![Confusion_matrix](./images/confusion_matrix.png)

### References
1. List of ICD-9 codes : https://en.wikipedia.org/wiki/List_of_ICD-9_codes
2. Optuna documentation : https://optuna.readthedocs.io/en/stable/
3. Pandas documentation : https://pandas.pydata.org/docs/
4. Numpy documentation : https://numpy.org/doc/1.26/user/index.html
5. Matplotlib.pyplot documentation : https://matplotlib.org/stable/users/index.html
6. Scikit learn documentation : https://scikit-learn.org/stable/
7. Seaborn documentation : https://seaborn.pydata.org/tutorial.html
8. Plotly documentation : https://plotly.com/python/plotly-fundamentals/
